src/Catalyst.Core.Modules.Kvm/DeltaExecutor.cs

```diff
@@ -0,0 +1,519 @@
+        ///     in the business sense. <see cref="IStorageProvider" /> needs to handle storage trees for various accounts
+        ///     while <see cref="IStateProvider" /> handles the basic accounts state with storage roots only.
+        /// </summary>
+        /// <param name="specProvider">The network upgrade spec - defines the virtual machine version.</param>
+        /// <param name="stateProvider">Access to accounts.</param>
+        /// <param name="storageProvider">Access to accounts' storage.</param>
+        /// <param name="virtualMachine">A virtual machine to execute the code on.</param>
+        /// <param name="logger">Logger for the execution details.</param>
+        public DeltaExecutor(ISpecProvider specProvider,
+            IStateProvider stateProvider,
+            IStorageProvider storageProvider,
+            IVirtualMachine virtualMachine,
+            ILogger logger)
+        {
+            _logger = logger;
+            _specProvider = specProvider;
+            _virtualMachine = virtualMachine;
+            _stateProvider = stateProvider;
+            _storageProvider = storageProvider;
+        }
+
+        [Todo("Wider work needed to split calls and execution properly")]
+        public void CallAndRestore(Delta stateUpdate, ITxTracer txTracer) { Execute(stateUpdate, txTracer, true); }
+
+        [Todo("After delta is executed we should validate the state root and if it is not as expected we should revert all the changes.")]
+        public void Execute(Delta stateUpdate, ITxTracer txTracer) { Execute(stateUpdate, txTracer, false); }
+
+        /// <summary>
+        ///     This method calculates the basic gas cost of the <paramref name="entry" />
+        ///     (excluding the cost of the VM execution).
+        ///     The intrinsic cost is calculated as 21000 base cost + the entry data cost that is a function
+        ///     of the number of bytes of data (and the cost of a single byte may differ between 0 bytes and non-0 bytes.
+        /// </summary>
+        /// <param name="entry"></param>
+        /// <param name="releaseSpec"></param>
+        /// <returns>Total intrinsic cost of the <paramref name="entry" /></returns>
+        public ulong CalculateIntrinsicGas(ContractEntry entry, IReleaseSpec releaseSpec)
+        {
+            ulong result = GasCostOf.Transaction; // the basic entry cost
+            if (entry.Data != null)
+            {
+                // here is the difference between the 0 bytes and non-zero bytes cost
+                // justified by a better compression level of zero bytes
+                long txDataNonZeroGasCost = releaseSpec.IsEip2028Enabled ? GasCostOf.TxDataNonZeroEip2028 : GasCostOf.TxDataNonZero;
+                int length = entry.Data.Length;
+                for (int i = 0; i < length; i++) result += entry.Data[i] == 0 ? GasCostOf.TxDataZero : (ulong) txDataNonZeroGasCost;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'for' block.](https://app.codacy.com/app/catalyst-network/Catalyst.Node/pullRequest?prid=4350015)